[draft] Add tutorial inductor on Windows CPU (#3062)
pytorch/tutorials: prototype_source/inductor_windows_cpu.rst

How to use TorchInductor on Windows CPU
=======================================

**Author**: `Zhaoqiong Zheng <https://github.com/ZhaoqiongZ>`_, `Xu, Han <https://github.com/xuhancn>`_

TorchInductor is a new compiler backend that compiles FX Graphs generated by TorchDynamo into optimized C++/Triton kernels.
This tutorial will guide you through the process of using TorchInductor on a Windows CPU.

.. grid:: 2

    .. grid-item-card:: :octicon:`mortar-board;1em;` What you will learn
       :class-card: card-prerequisites

       * How to compile and execute a Python function with PyTorch, optimized for Windows CPU
       * Basics of TorchInductor's optimization using C++/Triton kernels.

    .. grid-item-card:: :octicon:`list-unordered;1em;` Prerequisites
       :class-card: card-prerequisites

       * PyTorch v2.5 or later
       * Microsoft Visual C++ (MSVC)
       * Miniforge for Windows

Install the Required Software

> ---------------------

Reviewer: ```suggestion
-------------------------------------------
```

First, let's install the required software. C++ compiler is required for TorchInductor optimization.
We will use Microsoft Visual C++ (MSVC) for this example. 

1. Download and install `MSVC <https://visualstudio.microsoft.com/downloads/>`_.

> 1. During the installation, choose **Desktop Development with C++** in the **Desktop & Mobile** section. Then install the software

Reviewer: ```suggestion
2. During the installation, choose **Desktop Development with C++** in the **Desktop & Mobile** section. Then install the software
```

>   .. note::

Reviewer: ```suggestion
.. note::
```

     We recommend C++ compiler `Clang <https://github.com/llvm/llvm-project/releases>`_ and `Intel Compiler <https://www.intel.com/content/www/us/en/developer/tools/oneapi/base-toolkit-download.html>`_.

> 1. Download and install `Miniforge3-Windows-x86_64.exe <https://github.com/conda-forge/miniforge/releases/latest/>`__.

Reviewer: ```suggestion
3. Download and install `Miniforge3-Windows-x86_64.exe <https://github.com/conda-forge/miniforge/releases/latest/>`__.
```

> Set Up Environment
> ^^^^^^^^^^^^^^^^^^

Reviewer: ```suggestion
Set Up the Environment
------------------------------
```

#. Open the command line environment via ``cmd.exe``.
#. Activate `MSVC` with the following command::

    "C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Auxiliary/Build/vcvars64.bat"
#. Activate `conda` with the following command:

   .. code-block:: sh

    "C:/ProgramData/miniforge3/Scripts/activate.bat"
#. Create and activate a customer conda environment:
 

>      .. code-block:: sh
> 
>     conda create -n inductor_cpu_windows python=3.10 -y 
>     conda activate inductor_cpu_windows

Reviewer: ```suggestion
    .. code-block:: sh

      conda create -n inductor_cpu_windows python=3.10 -y 
      conda activate inductor_cpu_windows
```

How to use TorchInductor on Windows CPU
=======================================

**Author**: `Zhaoqiong Zheng <https://github.com/ZhaoqiongZ>`_, `Xu, Han <https://github.com/xuhancn>`_

> TorchInductor is a new compiler backend that compiles FX Graphs generated by TorchDynamo into optimized C++/Triton kernels.

Reviewer: We should avoid using words like "new" as it gets outdated quickly:
Also, saying "compiler compiles" might be redundant, can we say "transform" instead?
```suggestion
TorchInductor is a compiler backend that transforms FX Graphs generated by TorchDynamo into highly optimized C++/Triton kernels.
```

            6.1945e-01,  4.6654e-01, -3.7376e-01,  9.3644e-01,  1.3600e+00],
            [-1.0157e-01,  7.7200e-02,  1.0146e+00,  8.8175e-02, -1.4057e+00,
            8.8119e-01,  6.2853e-01,  3.2773e-01,  8.5082e-01,  8.4615e-01],
            [ 1.4140e+00,  1.2130e+00, -2.0762e-01,  3.3914e-01,  4.1122e-01,
            8.6895e-01,  5.8852e-01,  9.3310e-01,  1.4101e+00,  9.8318e-01],
            [ 1.2355e+00,  7.9290e-02,  1.3707e+00,  1.3754e+00,  1.3768e+00,
            9.8970e-01,  1.1171e+00, -5.9944e-01,  1.2553e+00,  1.3394e+00],
            [-1.3428e+00,  1.8400e-01,  1.1756e+00, -3.0654e-01,  9.7973e-01,
            1.4019e+00,  1.1886e+00, -1.9194e-01,  1.3632e+00,  1.1811e+00],
            [-7.1615e-01,  4.6622e-01,  1.2089e+00,  9.2011e-01,  1.0659e+00,
            9.0892e-01,  1.1932e+00,  1.3888e+00,  1.3898e+00,  1.3218e+00],
            [ 1.4139e+00, -1.4000e-01,  9.1192e-01,  3.0175e-01, -9.6432e-01,
            -1.0498e+00,  1.4115e+00, -9.3212e-01, -9.0964e-01,  1.0127e+00],
            [ 5.7244e-04,  1.2799e+00,  1.3595e+00,  1.0907e+00,  3.7191e-01,
            1.4062e+00,  1.3672e+00,  6.8502e-02,  8.5216e-01,  8.6046e-01]])

> Alternative Compiler for better performance

Reviewer: ```suggestion
Using an Alternative Compiler for Better Performance
```

-------------------------------------------

> To enhance performance on Windows inductor, you can use the Intel Compiler or LLVM Compiler. However, they rely on the runtime libraries from Microsoft Visual C++ (MSVC). Therefore, your first step should be to install MSVC.

Reviewer 2: It would be good to show an improvement in perf for the same code above with the intel compiler & llvm compiler or at least talk about how these compilers are more performant

Reviewer 3: Hi @agunapal ,

I'm working with Intel compiler team to publish another blog, which shows detailed guide to run Windows inductor on Intel compiler environment. And it also contains performance data.
After that blog done, we will publish on Intel development community blog, and we will link to that blog from this doc.

Reviewer: @xuhancn can you create an issue in this repo for this follow up work?

Intel Compiler
^^^^^^^^^^^^^^

#. Download and install `Intel Compiler <https://www.intel.com/content/www/us/en/developer/tools/oneapi/dpc-compiler-download.html>`_ with Windows version.

> #. Set Windows Inductor Compiler via environment variable ``set CXX=icx-cl``

Reviewer: ```suggestion
#. Set Windows Inductor Compiler with the CXX environment variable ``set CXX=icx-cl``.
```

LLVM Compiler
^^^^^^^^^^^^^

#. Download and install `LLVM Compiler <https://github.com/llvm/llvm-project/releases>`_ and choose win64 version.

> #. Set Windows Inductor Compiler via environment variable ``set CXX=clang-cl`` 

Reviewer: ```suggestion
#. Set Windows Inductor Compiler with the CXX environment variable ``set CXX=clang-cl``.
```

Conclusion
----------

> With this tutorial, we introduce how to use Inductor on Windows CPU with PyTorch 2.5 or later. We can use Intel Compiler or LLVM Compiler to get better performance.

Reviewer: ```suggestion
In this tutorial, we have learned how to use Inductor on Windows CPU with PyTorch. In addition, we discussed
further performance improvements with Intel Compiler and LLVM Compiler.
```

We will use Microsoft Visual C++ (MSVC) for this example. 

1. Download and install `MSVC <https://visualstudio.microsoft.com/downloads/>`_.

2. During the installation, choose **Desktop Development with C++** in the **Desktop & Mobile** section in **Workloads** table. Then install the software

.. note::

     We recommend C++ compiler `Clang <https://github.com/llvm/llvm-project/releases>`_ and `Intel Compiler <https://www.intel.com/content/www/us/en/developer/tools/oneapi/base-toolkit-download.html>`_.
     Please check `Alternative Compiler for better performance <#alternative-compiler-for-better-performance>`_.

3. Download and install `Miniforge3-Windows-x86_64.exe <https://github.com/conda-forge/miniforge/releases/latest/>`__.

Set Up the Environment
----------------------

#. Open the command line environment via ``cmd.exe``.
#. Activate ``MSVC`` with the following command:

   .. code-block:: sh

    "C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Auxiliary/Build/vcvars64.bat"
#. Activate ``conda`` with the following command:

   .. code-block:: sh

    "C:/ProgramData/miniforge3/Scripts/activate.bat"

> #. Create and activate a customer conda environment:

Reviewer 4: It should be "custom" conda environment, shouldn't it?

 
   .. code-block:: sh

    conda create -n inductor_cpu_windows python=3.10 -y 
    conda activate inductor_cpu_windows

#. Install `PyTorch 2.5 <https://pytorch.org/get-started/locally/>`_ or later.

Using TorchInductor on Windows CPU
----------------------------------

Here’s a simple example to demonstrate how to use TorchInductor:

.. code-block:: python

    import torch
    def foo(x, y):
        a = torch.sin(x)

>         b = torch.cos(x)

Reviewer 4: If function takes two arguments, should 2nd one be used somewhere? (i.e. `y` argument is currently unused in the codebase, is it?)

        return a + b
    opt_foo1 = torch.compile(foo)
    print(opt_foo1(torch.randn(10, 10), torch.randn(10, 10)))

> The code above returns the following output: 

Reviewer 4: It would not, would it? As inputs are random

How to use TorchInductor on Windows CPU
=======================================

**Author**: `Zhaoqiong Zheng <https://github.com/ZhaoqiongZ>`_, `Xu, Han <https://github.com/xuhancn>`_

Introduction
------------

TorchInductor is the new compiler backend that compiles the FX Graphs generated by TorchDynamo into optimized C++/Triton kernels.

This tutorial introduces the steps for utilizing TorchInductor on Windows CPU.

Software Installation
---------------------

Now, we will walk you through a step-by-step tutorial for how to use torchinductor on Windows CPU.

> 
> Install a Compiler
> ^^^^^^^^^^^^^^^^^^
> 
> C++ compiler is required for torchinductor optimization, let's take Microsoft Visual C++ (MSVC) as an example.
> 
> Download and install `MSVC <https://visualstudio.microsoft.com/downloads/>`_.

Reviewer: ```suggestion

1. Download and install `MSVC <https://visualstudio.microsoft.com/downloads/>`_.
```

> During Installation, chosse `Desktop Development with C++` in the `Desktop & Mobile` Section and then install.

Reviewer: ```suggestion
1. During the installation, choose **Desktop Development with C++** in the **Desktop & Mobile** section. Then install the software
```

> .. note::
> 
>     We recommend C++ compiler `Clang <https://github.com/llvm/llvm-project/releases>`_ and `Intel Compiler <https://www.intel.com/content/www/us/en/developer/tools/oneapi/base-toolkit-download.html>`_.

Reviewer: ```suggestion
  .. note::

     We recommend C++ compiler `Clang <https://github.com/llvm/llvm-project/releases>`_ and `Intel Compiler <https://www.intel.com/content/www/us/en/developer/tools/oneapi/base-toolkit-download.html>`_.
```

> Install Miniforge
> ^^^^^^^^^^^^^^^^^
> 
> Download and install `Miniforge <https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-Windows-x86_64.exe>`_.

Reviewer: ```suggestion

1. Download and install `Miniforge3-Windows-x86_64.exe <https://github.com/conda-forge/miniforge/releases/latest/>`__.
```

Set Up Environment
^^^^^^^^^^^^^^^^^^

> #. Open a command line environment via cmd.exe.

Reviewer: ```suggestion
#. Open the command line environment via ``cmd.exe``.
```

> #. Activate `MSVC` via below command::

Reviewer: ```suggestion
#. Activate `MSVC` with the following command::
```

    "C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Auxiliary/Build/vcvars64.bat"

> #. Activate `conda` via below command::

Reviewer: ```suggestion
#. Activate `conda` with the following command:

   .. code-block:: sh
```

    "C:/ProgramData/miniforge3/Scripts/activate.bat"

> #. Create and activate customer conda environment::

Reviewer: ```suggestion
#. Create and activate a customer conda environment:
 
     .. code-block:: sh
```

    conda create -n inductor_cpu_windows python=3.10 -y 
    conda activate inductor_cpu_windows

> #. Install `PyTorch 2.5 <https://pytorch.org/get-started/locally/>`_ or later.

Reviewer: ```suggestion

#. Install `PyTorch 2.5 <https://pytorch.org/get-started/locally/>`_ or later.
```

> #. Use torchinductor on Windows CPU::

Reviewer: ```suggestion

Using TorchInductor on Windows CPU
----------------------------------------

Here’s a simple example to demonstrate how to use TorchInductor:

.. code-block:: python

```

    import torch
    def foo(x, y):
        a = torch.sin(x)
        b = torch.cos(x)
        return a + b
    opt_foo1 = torch.compile(foo)
    print(opt_foo1(torch.randn(10, 10), torch.randn(10, 10)))

> #. Output of the above example::

Reviewer: ```suggestion
The code above returns the following output: 

.. code-block:: sh
```

            6.1945e-01,  4.6654e-01, -3.7376e-01,  9.3644e-01,  1.3600e+00],
            [-1.0157e-01,  7.7200e-02,  1.0146e+00,  8.8175e-02, -1.4057e+00,
            8.8119e-01,  6.2853e-01,  3.2773e-01,  8.5082e-01,  8.4615e-01],
            [ 1.4140e+00,  1.2130e+00, -2.0762e-01,  3.3914e-01,  4.1122e-01,
            8.6895e-01,  5.8852e-01,  9.3310e-01,  1.4101e+00,  9.8318e-01],
            [ 1.2355e+00,  7.9290e-02,  1.3707e+00,  1.3754e+00,  1.3768e+00,
            9.8970e-01,  1.1171e+00, -5.9944e-01,  1.2553e+00,  1.3394e+00],
            [-1.3428e+00,  1.8400e-01,  1.1756e+00, -3.0654e-01,  9.7973e-01,
            1.4019e+00,  1.1886e+00, -1.9194e-01,  1.3632e+00,  1.1811e+00],
            [-7.1615e-01,  4.6622e-01,  1.2089e+00,  9.2011e-01,  1.0659e+00,
            9.0892e-01,  1.1932e+00,  1.3888e+00,  1.3898e+00,  1.3218e+00],
            [ 1.4139e+00, -1.4000e-01,  9.1192e-01,  3.0175e-01, -9.6432e-01,
            -1.0498e+00,  1.4115e+00, -9.3212e-01, -9.0964e-01,  1.0127e+00],
            [ 5.7244e-04,  1.2799e+00,  1.3595e+00,  1.0907e+00,  3.7191e-01,
            1.4062e+00,  1.3672e+00,  6.8502e-02,  8.5216e-01,  8.6046e-01]])

Conclusion
----------

> With this tutorial, we introduce how to use Inductor on Windows CPU with PyTorch 2.5 or later.  

Reviewer: ```suggestion
In this tutorial, we have learned how to use Inductor on Windows CPU with PyTorch 2.5 or later.
```

> With this tutorial, we introduce how to use Inductor on Windows CPU with PyTorch 2.5 or later.  

Reviewer: Any links we can add for user to explore more?

Author: It is just a simple guide to use the feature "inductor on Windows CPU".
I add an alternative compiler section for better performance with Intel Compiler or LLVM Compiler
